Make B3 Origin a tagged union

justinmichaud opened the pull request:
#### 1550e2947171f4c5a8bf47ad1536c38bdd101510
<pre>
Make B3 Origin a tagged union
<a href="https://bugs.webkit.org/show_bug.cgi?id=288575">https://bugs.webkit.org/show_bug.cgi?id=288575</a>

Reviewed by Keith Miller.

In <a href="https://commits.webkit.org/289952@main">https://commits.webkit.org/289952@main</a>, Yusuke introduced
a new way to handle OMG inline frame exceptions using code origins.

- We no longer allow origins to be imprecise on ARMv7
- We slightly reduce the number of extended wasm opcodes that can be
    packed to add a tag bit to OpcodeOrigin
- We use this tag bit in OMG to verify that we get the expected
    origin kind

At least on ARMv7, we can have both OMGOrigins and OpcodeOrigins,
so it remains to be investigated why this patch did not introduce
failures on 64-bit platforms.

* Source/JavaScriptCore/b3/B3Origin.cpp:
(JSC::B3::Origin::dump const):
* Source/JavaScriptCore/b3/B3Origin.h:
(JSC::B3::Origin::Origin):
(JSC::B3::Origin::isOMGOrigin const):
(JSC::B3::Origin::isPackedWasmOrigin const):
(JSC::B3::Origin::omgOrigin const):
(JSC::B3::Origin::maybeOMGOrigin const):
(JSC::B3::Origin::data const): Deleted.
* Source/JavaScriptCore/jit/PCToCodeOriginMap.h:
* Source/JavaScriptCore/wasm/WasmBBQJIT.h:
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::OptimizingJITCallee::materializePCToOriginMap):
* Source/JavaScriptCore/wasm/WasmCompilationContext.h:
(): Deleted.
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
(JSC::Wasm::OMGIRGenerator::emitExceptionCheck):
(JSC::Wasm::OMGIRGenerator::addThrow):
(JSC::Wasm::OMGIRGenerator::addThrowRef):
(JSC::Wasm::OMGIRGenerator::addRethrow):
(JSC::Wasm::parseAndCompileOMG):
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator32_64.cpp:
(JSC::Wasm::OMGIRGenerator::origin):
(JSC::Wasm::parseAndCompileOMG):
* Source/JavaScriptCore/wasm/WasmOpcodeOrigin.h:
(JSC::Wasm::OpcodeOrigin::OpcodeOrigin):
(JSC::Wasm::OpcodeOrigin::opcode const):
(JSC::Wasm::OpcodeOrigin::ext1Opcode const):
(JSC::Wasm::OpcodeOrigin::simdOpcode const):
(JSC::Wasm::OpcodeOrigin::gcOpcode const):
(JSC::Wasm::OpcodeOrigin::atomicOpcode const):
(JSC::Wasm::OpcodeOrigin::location const):
(JSC::Wasm::OpcodeOrigin::asB3Origin):

Canonical link: <a href="https://commits.webkit.org/292486@main">https://commits.webkit.org/292486@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e7e6a39265ae3fd375851a2b76b63cafd8870765

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95737 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15345 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5260 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/100788 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46242 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15633 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/23778 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73007 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30262 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98740 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11706 "Found 1 new test failure: fast/forms/ios/focus-input-via-button.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86465 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53341 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11431 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4192 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45580 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/88405 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81601 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4302 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/102822 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/94357 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/22795 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/16622 "Found 2 new test failures: html5lib/generated/run-entities01-write.html imported/w3c/web-platform-tests/html/infrastructure/safe-passing-of-structured-data/shared-array-buffers/no-coop-coep.https.any.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82049 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23047 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82491 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81412 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20517 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25984 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3441 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/16132 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/22763 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/27912 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/117836 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22422 "Built successfully") | | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/33366 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Found 319 jsc stress test failures: microbenchmarks/interpreter-wasm.js.default, microbenchmarks/memcpy-wasm-large.js.bytecode-cache, microbenchmarks/memcpy-wasm-large.js.default, microbenchmarks/memcpy-wasm-large.js.dfg-eager, microbenchmarks/memcpy-wasm-large.js.dfg-eager-no-cjit-validate ...; Compiled JSC; Found 328 jsc stress test failures: microbenchmarks/interpreter-wasm.js.default, microbenchmarks/memcpy-wasm-large.js.bytecode-cache, microbenchmarks/memcpy-wasm-large.js.default, microbenchmarks/memcpy-wasm-large.js.dfg-eager, microbenchmarks/memcpy-wasm-large.js.dfg-eager-no-cjit-validate ...; Found 7 new JSC stress test failures: microbenchmarks/memcpy-wasm-medium.js.bytecode-cache, wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.default-wasm, wasm.yaml/wasm/stress/ipint-bbq-osr-with-try5.js.wasm-eager, wasm.yaml/wasm/stress/live-funcref-across-loop-tier-up.js.default-wasm, wasm.yaml/wasm/stress/live-funcref-across-loop-tier-up.js.wasm-eager, wasm.yaml/wasm/stress/repro_1289.js.default-wasm, wasm.yaml/wasm/stress/simple-inline-exception-inlinee-catch-with-delegate.js.default-wasm (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/25898 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24164 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 reviewed (changes requested):
Why not just use `TaggedPtr<void*, EnumTaggingTraits<OriginType>>` with some `enum class OriginType { OMGOrigin, OpcodeOrigin };`? Seems like you'd reduce a lot of logic here.

kmiller68 commented on Source/WTF/wtf/StdLibExtras.h:
```diff
@@ -1540,6 +1540,47 @@ ALWAYS_INLINE void lazyInitialize(const std::unique_ptr<T>& ptr, std::unique_ptr
+    ASSERT(casted->magicValue == T::MAGIC_VALUE);
+    return casted;
+}
+
```
This is very close to the `downcast` in `TypeCasts.h`, which I think is going to be confusing for folks in the long run... It's also used as a validated reinterpret_cast between unrelated types rather than a downcast so maybe `ValidatedReinterpretCast` is a better name. This also seems a bit too specific to be part of `StdLibExtras.h`, can we it to its own header?

kmiller68 commented on Source/WTF/wtf/TaggedPtr.h:
```diff
@@ -115,6 +115,37 @@ struct NoTaggingTraits {
     static TagType unwrapTag(StorageType) { return defaultTag; }
 };
 
+class Bits60 {
```
This should probably be `TaggedBits60` or something.

kmiller68 reviewed (approved):
r=me.

justinmichaud committed:
Make B3 Origin a tagged union
https://bugs.webkit.org/show_bug.cgi?id=288575

Reviewed by Keith Miller.

In https://commits.webkit.org/289952@main, Yusuke introduced
a new way to handle OMG inline frame exceptions using code origins.

- We no longer allow origins to be imprecise on ARMv7
- We slightly reduce the number of extended wasm opcodes that can be
    packed to add a tag bit to OpcodeOrigin
- We use this tag bit in OMG to verify that we get the expected
    origin kind

At least on ARMv7, we can have both OMGOrigins and OpcodeOrigins,
so it remains to be investigated why this patch did not introduce
failures on 64-bit platforms.

* Source/JavaScriptCore/b3/B3Origin.cpp:
(JSC::B3::Origin::dump const):
* Source/JavaScriptCore/b3/B3Origin.h:
(JSC::B3::Origin::Origin):
(JSC::B3::Origin::isOMGOrigin const):
(JSC::B3::Origin::isPackedWasmOrigin const):
(JSC::B3::Origin::omgOrigin const):
(JSC::B3::Origin::maybeOMGOrigin const):
(JSC::B3::Origin::data const): Deleted.
* Source/JavaScriptCore/jit/PCToCodeOriginMap.h:
* Source/JavaScriptCore/wasm/WasmBBQJIT.h:
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::OptimizingJITCallee::materializePCToOriginMap):
* Source/JavaScriptCore/wasm/WasmCompilationContext.h:
(): Deleted.
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
(JSC::Wasm::OMGIRGenerator::emitExceptionCheck):
(JSC::Wasm::OMGIRGenerator::addThrow):
(JSC::Wasm::OMGIRGenerator::addThrowRef):
(JSC::Wasm::OMGIRGenerator::addRethrow):
(JSC::Wasm::parseAndCompileOMG):
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator32_64.cpp:
(JSC::Wasm::OMGIRGenerator::origin):
(JSC::Wasm::parseAndCompileOMG):
* Source/JavaScriptCore/wasm/WasmOpcodeOrigin.h:
(JSC::Wasm::OpcodeOrigin::OpcodeOrigin):
(JSC::Wasm::OpcodeOrigin::opcode const):
(JSC::Wasm::OpcodeOrigin::ext1Opcode const):
(JSC::Wasm::OpcodeOrigin::simdOpcode const):
(JSC::Wasm::OpcodeOrigin::gcOpcode const):
(JSC::Wasm::OpcodeOrigin::atomicOpcode const):
(JSC::Wasm::OpcodeOrigin::location const):
(JSC::Wasm::OpcodeOrigin::asB3Origin):

Canonical link: https://commits.webkit.org/292486@main
